### ISSUE TITLE
Add support for action or script exec when installing sourceless routes

### ISSUE THREAD

**Issue by troglobit** [troglobit/smcroute#14]
I recently tried to set up a sourceless (*,G) multicast route in SMCRoute on a system
that also did 1-to-1 NAT of the source address. This didn't work since conntrack locked
on to the un-natted source IP when SMCRoute installed the source specific (S,G) route
in the kernel upon receiving the first frame from a sender.

         .-------.  ETB  .-------.    ETB    .-------.  ETB  .-------.
         |       |=======|       |===========|       |=======|       |
         |  TR1  |       |  TR2  |           |  TR3  |       |  TR4  |
         |       |-------|       |           |       |-------|       |
         `-------'  ECN  `-------'           `-------'  ECN  `-------'
             |               |                   |               |
         ECN |  10.0.0.0/24  | ECN           ECN |  10.0.0.0/24  | ECN
             |               |                   |               |
         .-------.  ECN  .-------.           .-------.  ECN  .-------.
         |       |-------|       |           |       |-------|       |
         |  RS1  |       |  RS2  |           |  RS3  |       |  RS4  |
         |       |       |       |           |       |       |       |
         `-------'       `-------'           `-------'       `-------'
                  Car #1                              Car #2

There are of course ways around this using iptables, but in the spirit of keeping things
simple, the following approach may be the best.

If we add an action `conntrack-flush` and/or support for calling an external script when
installing the source specific route this problem can be avoided.  It could look something
like this:

    mroute from eth0 group 225.3.2.1 to eth1 eth2 [exec /path/to/script | conntrack-flush]

**Comment by bombadil** [comment]
This seems to be an interesting idea. Yet I don't fully get how this is intended to solve your issue.

Can you please explain a bit more verbose:
1. What do you refer to when using the term "installing the source specific route", i.e. what exactly would be the trigger?
2. When exactly would the conntrack-flush or script action get triggered?
3. How does this action solve the issue you're trying to solve?

The example you provided at the end looks like an entry in smcroute.conf. Is it really a good idea to put the definition of trigger actions in the same line? In the spirit of keeping things simple, maybe the actions can be defined independent of specific routes, i.e. depending on certain global trigger events. This would allow to put the action definition on a separate line.

**Comment by troglobit** [comment]
Hi Micha :)

  1. When we recieve an `IGMPMSG_NOCACHE` message we call `mroute4_dyn_add()` to check for a matching (*,G) route to add to the kernel
  2. When this function returns OK it means we've installed a new multicast route and we can call an external script
  3. It's not until a multicast route is installed in the kernel that the NAT 1-to-1 rule will bite, so depending on how that rule looks the traffic may be dropped or NAT:ed wrong. With a callback we can fix this situation by flushing the connection tracking in netfilter.

Agreed, a global action could also be used.  Just wondering what the semantics of that would be:

* Call action after any mroute is installed?
* Call action after all mroutes have been installed?
* Call action only for "dynamic" (source-less) mroutes?

    mroute-install-action <exec /path/to/script | conntrack-flush>

**Comment by bombadil** [comment]
I agree that the semantics wouldn't be clear if you have only one global action. However I would like to see the configuration of /when/ to trigger a conntrack-flush separated from /how/ to trigger it.

What do you think about an optional flag "conntrack-flush" at the end of the route specification, as you initially suggested:
```
mroute from eth0 group 225.3.2.1 to eth1 eth2 [conntrack-flush]
```
... and an independent optional setting (i.e. on a separate line) that specifies a script to execute when the conntrack-flush action is about to be (or just has been) performed. E.g.:
```
conntrack-flush-exec /some/script
```
This way you only need to specify the script once, even if you have multiple routes that should trigger the conntrack-flush.

**Comment by troglobit** [comment]
That's fair, separating *when* from *how* -- I like that!  I also like your idea of only having to specify
a script once.  But maybe one can make it even more generic?  I also started thinking about having
separate pre/post actions, i.e. logging suspect multicast traffic in a pre-action and returning FALSE
to smcroute to deny installing it ...

... I have to think about this incredible potential some more.

Thanks for the feedack, and sorry for my late repy!

**Comment by troglobit** [comment]
@bombadil I finally got around to formalizing a patch for this, I've put it up for audit as pull request #20. If you have the time I'd be very happy to get some feedback.

In particular I'm worried about the security aspects of running a script as root upon reception of a single IP multicast frame.  Maybe some added caution would be in order?  E.g., checking permissions/ownership of the script before executing it ... I don't know.

**Comment by bombadil** [comment]
@troglobit I'll have time to take a look on the next weekend.

**Comment by troglobit** [comment]
@bombadil That would be greatly appreciated, thank you! :smiley:

**Comment by bombadil** [comment]
I've added a few comments to commit 96cde51.

**Comment by troglobit** [comment]
Thanks, quite a few comments ... I think I'll have to rework the patch completely and do another pull request later.

**Comment by troglobit** [comment]
In pull request #21 I've prepared another version of the same idea, based on the discussion in #20. Tried to keep this one simple and without any extra bells and whistles.

**Comment by bombadil** [comment]
This approach is much nicer and cleaner. Good work! Let's continue with more commits based on that.

**Comment by troglobit** [comment]
Thanks @bombadil, very nice (and patient! ;) of you to take the time to audit!

**Comment by troglobit** [comment]
Pushed to master as 21a8251.
